File: balanced_binary_tree/balanced_binary_tree.ts
// class TreeNode {
//   val: number;
//   left: TreeNode | null;
//   right: TreeNode | null;
//   constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
//     this.val = val === undefined ? 0 : val;
//     this.left = left === undefined ? null : left;
//     this.right = right === undefined ? null : right;
//   }
// }

const isBalanced = (root: TreeNode) => {
  const dfs = (node: TreeNode | null) => {
    if (!node) {
      return [true, 0];
    }
    const left = dfs(node.left);
    const right = dfs(node.right);
    if (left[0] && right[0] && Math.abs(left[1] - right[1]) <= 1) {
      return [true, 1 + Math.max(left[1], right[1])];
    } else {
      return [false];
    }
  };
  return dfs(root)[0];
};

File: balanced_binary_tree/balanced_binary_tree.java
package balanced_binary_tree;

public class balanced_binary_tree {
    public class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;

        TreeNode() {
        }

        TreeNode(int val) {
            this.val = val;
        }

        TreeNode(int val, TreeNode left, TreeNode right) {
            this.val = val;
            this.left = left;
            this.right = right;
        }
    }

    public boolean isBalanced(TreeNode root) {
        return dfs(root).found;
    }

    public DfsResult dfs(TreeNode node) {
        if (node == null) {
            return new DfsResult(true, 0);
        }
        DfsResult left = dfs(node.left);
        DfsResult right = dfs(node.right);
        if (left.found && right.found && Math.abs(left.value - right.value) <= 1) {
            return new DfsResult(true, 1 + Math.max(left.value, right.value));
        } else {
            return new DfsResult(false, 0);
        }
    }

    public class DfsResult {
        public boolean found;
        public int value;

        public DfsResult(boolean found, int value) {
            this.found = found;
            this.value = value;
        }
    }
}
